opal_client/data/updater.py

```diff
@@ -233,5 +203,83 @@ async def wait_until_done(self):
         if self._subscriber_task is not None:
             await self._subscriber_task
 
+    @staticmethod
+    def calc_hash(data):
+        if not isinstance(data, str):
+            data = json.dumps(data)
+        return hashlib.sha256(data.encode('utf-8')).hexdigest()
+
+    async def report_update_results(self, update: DataUpdate, reports: List[DataEntryReport], data_fetcher: DataFetcher):
+        try:
+            whole_report = DataUpdateReport(update_id=update.id, reports=reports)
 
+            callbacks = update.callback.callbacks or opal_client_config.DEFAULT_UPDATE_CALLBACKS.callbacks
+            urls = []
+            for callback in callbacks:
+                if isinstance(callback, str):
+                    url = callback
+                    callback_config = opal_client_config.DEFAULT_UPDATE_CALLBACK_CONFIG.copy()
+                else:
+                    url, callback_config = callback
+                callback_config.data = whole_report.json()
+                urls.append((url, callback_config))
+
+            logger.info("Reporting the update to requested callbacks", urls=urls)
+            await data_fetcher.handle_urls(urls)
+            # TODO: log failed reports
+        except:
+            logger.exception("Failed to excute report_update_results")
+
+    async def update_policy_data(self, update: DataUpdate = None, policy_store: BasePolicyStoreClient = None, data_fetcher=None):
+        """
+        fetches policy data (policy configuration) from backend and updates it into policy-store (i.e. OPA)
+        """
+        policy_store = policy_store or DEFAULT_POLICY_STORE_GETTER()
+        if data_fetcher is None:
+            data_fetcher = DataFetcher()
+        # types / defaults
+        urls: List[Tuple[str, FetcherConfig]] = None
+        entries: List[DataSourceEntry] = []
+        # track the result of each url in order to report back
+        reports: List[DataEntryReport] = []
+        # if we have an actual specification for the update
+        if update is not None:
+            entries = update.entries
+            urls = [(entry.url, entry.config) for entry in entries]
 
+        # get the data for the update
+        logger.info("Fetching policy data", urls=urls)
+        # Urls may be None - handle_urls has a default for None
+        policy_data_with_urls = await data_fetcher.handle_urls(urls)
+        # Save the data from the update
+        # We wrap our interaction with the policy store with a transaction  
+        async with policy_store:
```

Comment:
i don't like the solution with context manager. in general it's fine, but:
- on line 279 you "swallow" exceptions if something goes wrong, so "end_trascation" will mark success however opa is not "truly" ready since no data was written

if you would instead raise the exception, it's also not good because our current practice is to log and report and this will be a breaking change.

i suggest instead:
- don't use a context manager here
- instead, if one entry is failed -> mark a boolean "encountered_failure"
- at the end of the loop of entries -> if not encountered_failure -> call policy_story.mark_update_complete / mark_transaction_complete.

i am open to other suggestions

Reply:
Context managers are the right way to go about these issues in Python (its the only way to ensure that the transaction closes) - its the Pythonic way to do RAII. Otherwise there's a significant risk of resource leaks and unclosed transactions 
If there are specific exceptions you think should be handled / or propagated - let's handle those specifically.